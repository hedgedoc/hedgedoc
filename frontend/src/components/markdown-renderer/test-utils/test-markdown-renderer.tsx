/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { StoreProvider } from '../../../redux/store-provider'
import type { MarkdownRendererExtension } from '../extensions/_base-classes/markdown-renderer-extension'
import { MarkdownToReact } from '../markdown-to-react/markdown-to-react'
import React, { useMemo } from 'react'

export interface SimpleMarkdownRendererProps {
  content: string
  extensions: MarkdownRendererExtension[]
}

/**
 * A markdown renderer for tests.
 *
 * @param content The content to be rendered.
 * @param extensions The {@link MarkdownRendererExtension MarkdownExtensions} to use for rendering.
 */
export const TestMarkdownRenderer: React.FC<SimpleMarkdownRendererProps> = ({ content, extensions }) => {
  const lines = useMemo(() => content.split('\n'), [content])
  return (
    <StoreProvider>
      <MarkdownToReact
        markdownContentLines={lines}
        markdownRenderExtensions={extensions}
        newlinesAreBreaks={true}
        allowHtml={false}
      />
    </StoreProvider>
  )
}
