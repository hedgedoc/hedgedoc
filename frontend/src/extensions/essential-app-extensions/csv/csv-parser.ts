/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

/**
 * Parses a given text as comma separated values (CSV).
 *
 * @param csvText The raw csv text
 * @param csvColumnDelimiter The delimiter for the columns
 * @return the values splitted by rows and columns
 */
export const parseCsv = (csvText: string, csvColumnDelimiter: string): string[][] => {
  const rows = csvText.split('\n')
  if (!rows || rows.length === 0) {
    return []
  }
  const splitRegex = new RegExp(`${escapeRegexCharacters(csvColumnDelimiter)}(?=(?:[^"]*"[^"]*")*[^"]*$)`)
  return rows.filter((row) => row !== '').map((row) => row.split(splitRegex))
}

/**
 * Escapes regex characters in the given string, so it can be used as literal string in another regex.
 *
 * @param unsafe The unescaped string
 * @return The escaped string
 */
const escapeRegexCharacters = (unsafe: string): string => {
  return unsafe.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}
