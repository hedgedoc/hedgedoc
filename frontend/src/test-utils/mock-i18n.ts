/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { TFunction } from 'i18next'
import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'

/**
 * Initializes i18n with minimal settings and without any data, so it just returns the used key as translation.
 *
 * @return A promise that resolves if i18n has been initialized
 */
export const mockI18n = (): Promise<TFunction> => {
  return i18n.use(initReactI18next).init({
    lng: 'en',
    fallbackLng: 'en',
    ns: ['translationsNS'],
    defaultNS: 'translationsNS',
    interpolation: {
      escapeValue: false
    },
    resources: { en: { translationsNS: {} } }
  })
}
