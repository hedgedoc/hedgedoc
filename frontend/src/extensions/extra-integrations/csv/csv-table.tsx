/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { cypressId } from '../../../utils/cypress-attribute'
import { parseCsv } from './csv-parser'
import React, { useMemo } from 'react'

export interface CsvTableProps {
  code: string
  delimiter: string
  showHeader: boolean
  tableRowClassName?: string
  tableColumnClassName?: string
}

/**
 * Renders a csv table.
 *
 * @param code The csv code
 * @param delimiter The delimiter to be used
 * @param showHeader If the header should be shown.
 * @param tableRowClassName Additional class name for the rows.
 * @param tableColumnClassNameA Additional class name for the columns.
 */
export const CsvTable: React.FC<CsvTableProps> = ({
  code,
  delimiter,
  showHeader,
  tableRowClassName,
  tableColumnClassName
}) => {
  const { rowsWithColumns, headerRow } = useMemo(() => {
    const rowsWithColumns = parseCsv(code.trim(), delimiter)
    const headerRow = showHeader ? rowsWithColumns.splice(0, 1)[0] : []
    return { rowsWithColumns, headerRow }
  }, [code, delimiter, showHeader])

  const renderTableHeader = useMemo(() => {
    return headerRow.length === 0 ? undefined : (
      <thead>
        <tr>
          {headerRow.map((column, columnNumber) => (
            <th key={`header-${columnNumber}`}>{column}</th>
          ))}
        </tr>
      </thead>
    )
  }, [headerRow])

  const renderTableBody = useMemo(
    () => (
      <tbody>
        {rowsWithColumns.map((row, rowNumber) => (
          <tr className={tableRowClassName} key={`row-${rowNumber}`}>
            {row.map((column, columnIndex) => (
              <td className={tableColumnClassName} key={`cell-${rowNumber}-${columnIndex}`}>
                {column.replace(/^"|"$/g, '')}
              </td>
            ))}
          </tr>
        ))}
      </tbody>
    ),
    [rowsWithColumns, tableColumnClassName, tableRowClassName]
  )

  return (
    <table className={'csv-html-table table-striped'} {...cypressId('csv-html-table')}>
      {renderTableHeader}
      {renderTableBody}
    </table>
  )
}
