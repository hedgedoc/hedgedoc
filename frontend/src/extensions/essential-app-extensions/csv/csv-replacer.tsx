/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { CodeBlockComponentReplacer } from '../../../components/markdown-renderer/replace-components/code-block-component-replacer'
import type { NodeReplacement } from '../../../components/markdown-renderer/replace-components/component-replacer'
import {
  ComponentReplacer,
  DO_NOT_REPLACE
} from '../../../components/markdown-renderer/replace-components/component-replacer'
import { CsvTable } from './csv-table'
import type { Element } from 'domhandler'
import React from 'react'

/**
 * Detects code blocks with "csv" as language and renders them as table.
 */
export class CsvReplacer extends ComponentReplacer {
  public replace(codeNode: Element): NodeReplacement {
    const code = CodeBlockComponentReplacer.extractTextFromCodeNode(codeNode, 'csv')
    if (!code) {
      return DO_NOT_REPLACE
    }

    const extraData = codeNode.attribs['data-extra']
    const extraRegex = /\s*(delimiter=([^\s]*))?\s*(header)?/
    const extraInfos = extraRegex.exec(extraData)

    const delimiter = extraInfos?.[2] ?? ','
    const showHeader = extraInfos?.[3] !== undefined

    return <CsvTable code={code} delimiter={delimiter} showHeader={showHeader} />
  }
}
