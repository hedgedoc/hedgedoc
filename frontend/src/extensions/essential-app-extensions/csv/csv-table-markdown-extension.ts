/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/_base-classes/markdown-renderer-extension'
import type { ComponentReplacer } from '../../../components/markdown-renderer/replace-components/component-replacer'
import { CsvReplacer } from './csv-replacer'

/**
 * Adds support for csv tables to the markdown rendering using code fences with "csv" as language.
 */
export class CsvTableMarkdownExtension extends MarkdownRendererExtension {
  public buildReplacers(): ComponentReplacer[] {
    return [new CsvReplacer()]
  }
}
