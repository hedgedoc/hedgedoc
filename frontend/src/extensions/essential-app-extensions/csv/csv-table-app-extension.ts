/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { CheatsheetExtension } from '../../../components/editor-page/cheatsheet/cheatsheet-extension'
import {
  basicCompletion,
  codeFenceRegex
} from '../../../components/editor-page/editor-pane/autocompletions/basic-completion'
import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/_base-classes/markdown-renderer-extension'
import { AppExtension } from '../../_base-classes/app-extension'
import { CsvTableMarkdownExtension } from './csv-table-markdown-extension'
import type { CompletionSource } from '@codemirror/autocomplete'

/**
 * Adds support for csv tables to the markdown rendering.
 */
export class CsvTableAppExtension extends AppExtension {
  buildMarkdownRendererExtensions(): MarkdownRendererExtension[] {
    return [new CsvTableMarkdownExtension()]
  }

  buildCheatsheetExtensions(): CheatsheetExtension[] {
    return [{ i18nKey: 'csv', entries: [{ i18nKey: 'table' }, { i18nKey: 'header' }] }]
  }

  buildAutocompletion(): CompletionSource[] {
    return [basicCompletion(codeFenceRegex, '```csv delimiter=;\n\n```')]
  }
}
