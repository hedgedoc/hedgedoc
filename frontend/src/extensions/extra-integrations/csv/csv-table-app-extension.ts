/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { CheatsheetExtension } from '../../../components/editor-page/cheatsheet/cheatsheet-extension'
import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
import { AppExtension } from '../../base/app-extension'
import { CsvTableMarkdownExtension } from './csv-table-markdown-extension'

/**
 * Adds support for csv tables to the markdown rendering.
 */
export class CsvTableAppExtension extends AppExtension {
  buildMarkdownRendererExtensions(): MarkdownRendererExtension[] {
    return [new CsvTableMarkdownExtension()]
  }

  buildCheatsheetExtensions(): CheatsheetExtension[] {
    return [{ i18nKey: 'csv', entries: [{ i18nKey: 'table' }, { i18nKey: 'header' }] }]
  }
}
