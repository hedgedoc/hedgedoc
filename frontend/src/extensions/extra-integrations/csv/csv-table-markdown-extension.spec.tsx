/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { CodeProps } from '../../../components/markdown-renderer/replace-components/code-block-component-replacer'
import { TestMarkdownRenderer } from '../../../components/markdown-renderer/test-utils/test-markdown-renderer'
import { mockI18n } from '../../../test-utils/mock-i18n'
import * as CsvTableModule from '../csv/csv-table'
import { CsvTableMarkdownExtension } from './csv-table-markdown-extension'
import { render } from '@testing-library/react'
import React from 'react'

jest.mock('../csv/csv-table')

describe('CSV Table Markdown Extension', () => {
  beforeAll(async () => {
    jest.spyOn(CsvTableModule, 'CsvTable').mockImplementation((({ code }) => {
      return (
        <span>
          this is a mock for csv frame
          <code>{code}</code>
        </span>
      )
    }) as React.FC<CodeProps>)
    await mockI18n()
  })

  afterAll(() => {
    jest.resetModules()
    jest.restoreAllMocks()
  })

  it('renders a csv codeblock', () => {
    const view = render(
      <TestMarkdownRenderer extensions={[new CsvTableMarkdownExtension()]} content={'```csv\na;b;c\nd;e;f\n```'} />
    )
    expect(view.container).toMatchSnapshot()
  })
})
