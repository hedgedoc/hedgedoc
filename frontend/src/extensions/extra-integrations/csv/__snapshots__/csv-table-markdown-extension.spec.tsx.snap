// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSV Table Markdown Extension renders a csv codeblock 1`] = `
<div>
  <pre>
    <code
      class="csv"
    >
      a;b;c
d;e;f

    </code>
  </pre>
  

</div>
`;
